.github/workflows/check-changelog.yml

```diff
@@ -0,0 +1,42 @@
+name: Check Changelog
+# This check makes sure that the changelog is properly updated
+# when a PR introduces a change in a test file.
+# To bypass this check, label the PR with "No Changelog Needed".
+on:
+  pull_request:
+
+jobs:
+  check:
+    runs-on: ubuntu-latest
+    if: ${{ contains(github.event.pull_request.labels.*.name, 'No Changelog Needed') == 0 }}
+    steps:
+      - name: Get PR number and milestone
+        run: |
+          echo "PR_NUMBER=${{ github.event.pull_request.number }}" >> $GITHUB_ENV
+          echo "MILESTONE_TITLE=${{ github.event.pull_request.milestone.title }} >> $GITHUB_ENV
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: '0'
+      - name: Check the changelog only if tests have been modified
+        run: |
+          set -xe
+          tests=$(git diff --name-only origin/master)
+          if [[ "$tests" =~ tests ]]
+          then
+            result=$(cat ./doc/whats_new/v*.rst)
+            if [[ "$result" =~ :pr:\`$PR_NUMBER\` ]]
+            then
+              echo "Changelog has been updated."
+              check_mlstn=$(cat ./doc/whats_new/v${MILESTONE_TITLE:0:4}.rst)
+              if [[ "$check_mlstn" =~ :pr:\`$PR_NUMBER\` ]]
```

Comment:
I'm not sure what happens if there's no mileston label on the PR?

instead of checking all changelogs first and then the specific changelog, could we just check

- all changelogs if there's no label
- the expected changelog if there's a label?

Reply:
I tend to prefer the current solution for the following reason: I have seen PRs mile-stoned for the r+2 release while r+1 was under development.
Assuming that the changelog entry has been written in r+1
  - current implementation: you will be notified that the entry has been written but there is a mis-versioning
  - your suggestion: the check will fail. You should verify yourself if the entry is there.

I'm assuming that the furthest milestone has been assigned to notify low priority or to not put pressure ... I don't know... but as I've seen this quite often in the issue tracker I'm not comfortable in telling to people "don't do that" as long as an automatic check is affordable.

Reply:
Not at all... sorry, the furthest changelog is not present.

Reply:
>but the check will be green anyway

I'm a bit confused because as far as I understand the current check, it will fail if the entry isn't written in the corresponding changelog. Which is why it seems unnecessary to check all changelogs first

Reply:
I'm affected by a confirmation bias here ... :(
Another quite common case is an old milestone in a new PR:
- current implementation: the entry is there, the check will tell you that the milestone is wrong
- your suggestion: there is no new entry in the old changelog.

Again, I know that those cases should not happen, but they happen...

Reply:
IIUC your point is that the current version will provide a more informative error message? Fair point. I still find the proposed structure below easier to follow (and more efficient, which doesn't really matter anyway). Maybe we can just amend the general error message saying 

> This PR is modifying tests and we thus expect a changelog entry. If no changelog entry is needed, label the PR with `no changelog needed`. If you see this error and there is already a changelog entry, there may be a mismatch between the PR milestone and the changelog?

WDYT?